crates/lsp/src/lib.rs

```diff
@@ -142,22 +157,99 @@ fn on_did_close_text_document(
     ControlFlow::Continue(())
 }
 
+fn on_did_save_text_document(
+    state: &mut LspState,
+    params: DidSaveTextDocumentParams,
+) -> ControlFlow<Result<(), async_lsp::Error>> {
+    // TODO: Requiring `Language` and `is_opcode_supported` to construct a driver makes for some real stinky code
+    // The driver should not require knowledge of the backend; instead should be implemented as an independent pass (in nargo?)
+    let mut driver = Driver::new(&Language::R1CS, Box::new(|_op| false));
+
+    let file_path = &params.text_document.uri.to_file_path().unwrap();
+
+    driver.create_local_crate(file_path, CrateType::Binary);
+
+    let mut diagnostics = Vec::new();
+
+    let file_diagnostics = match driver.check_crate(false) {
+        Ok(warnings) => warnings,
+        Err(errors_and_warnings) => errors_and_warnings,
+    };
+
+    if !file_diagnostics.is_empty() {
+        let fm = driver.file_manager();
+        let files = fm.as_simple_files();
+
+        for FileDiagnostic { file_id, diagnostic } in file_diagnostics {
+            // TODO: This file_id never be 0 because the "path" where it maps is the directory, not a file
+            if file_id.as_usize() != 0 {
+                continue;
+            }
+
+            let mut range = Range::default();
+
+            // TODO: Should this be the first item in secondaries? Should we bail when we find a range?
+            for sec in diagnostic.secondaries {
+                // TODO: Codespan ranges are often (always?) off by some amount of characters
+                if let Ok(codespan_range) =
+                    codespan_lsp::byte_span_to_range(files, file_id.as_usize(), sec.span.into())
+                {
+                    // We have to manually attach each because the codespan_lsp restricts lsp-types to the wrong version range
+                    range.start.line = codespan_range.start.line;
+                    range.start.character = codespan_range.start.character;
+                    range.end.line = codespan_range.end.line;
+                    range.end.character = codespan_range.end.character;
+                }
+            }
+            let severity = match diagnostic.kind {
+                DiagnosticKind::Error => Some(DiagnosticSeverity::ERROR),
+                DiagnosticKind::Warning => Some(DiagnosticSeverity::WARNING),
+            };
```

Comment:
We'll probably want conversion functions for these. Either via implementing From or I'm fine with a free function as well since we're in another crate.

Reply:
Yeah, I didn't want to make the noirc_errors package depend on lsp-types.

I don't see the usefulness of making a function in this file for a small pattern match like this.

Reply:
I see the usefulness in functions because they are smaller and communicate intent better. This being 2 lines instead of 10 also helps a good deal in getting some of the less relevant conversion code out of the way when debugging the rest of the function. I also don't think this would be the only place in the future where we have to convert spans and diagnostic kinds.